tests/scripts/generate_bignum_tests.py

```diff
@@ -0,0 +1,299 @@
+    title = ""
+
+    def __init__(self) -> None:
+        type(self).count += 1
+
+    @property
+    def args(self) -> List[str]:
+        """Create list of arguments for test case."""
+        return []
+
+    @property
+    def description(self) -> str:
+        """Create a numbered test description."""
+        return "{} #{} {}".format(self.title, self.count, self.desc)
+
+    def create_test_case(self) -> test_case.TestCase:
+        """Generate test case from the current object."""
+        tc = test_case.TestCase()
+        tc.set_description(self.description)
+        tc.set_function(self.func)
+        tc.set_arguments(self.args)
+
+        return tc
+
+    @classmethod
+    def generate_tests(cls):
+        """Generate test cases for the target subclasses."""
+        for subclass in sorted(cls.__subclasses__(), key=lambda c: c.__name__):
+            yield from subclass.generate_tests()
+
+
+class BignumTarget(BaseTarget):
+    """Target for bignum (mpi) test case generation."""
+    gen_file = 'test_suite_mpi.generated'
+
+
+class BignumOperation(BignumTarget):
+    """Common features for test cases covering bignum operations.
+
+    Attributes:
+        symb: Symbol used for operation in description.
+        input_vals: List of values used to generate test case args.
+        input_cases: List of tuples containing test case inputs. This
+            can be used to implement specific pairs of inputs.
+    """
+    symb = ""
+    input_vals = [
+        "", "0", "7b", "-7b",
+        "0000000000000000123", "-0000000000000000123",
+        "1230000000000000000", "-1230000000000000000"
+    ] # type: List[str]
+    input_cases = [] # type: List[Tuple[str, ...]]
+
+    def __init__(self, val_l: str, val_r: str) -> None:
+        super().__init__()
+
+        self.arg_l = val_l
+        self.arg_r = val_r
+        self.int_l = hex_to_int(val_l)
+        self.int_r = hex_to_int(val_r)
+
+    @property
+    def args(self):
+        return [quote_str(self.arg_l), quote_str(self.arg_r), self.result]
+
+    @property
+    def description(self):
+        desc = self.desc if self.desc else "{} {} {}".format(
+            self.val_desc(self.arg_l),
+            self.symb,
+            self.val_desc(self.arg_r)
+        )
+        return "{} #{} {}".format(self.title, self.count, desc)
+
+    @property
+    def result(self) -> Optional[str]:
+        return None
+
+    @staticmethod
+    def val_desc(val) -> str:
+        """Generate description of the argument val."""
+        if val == "":
+            return "0 (null)"
+        if val == "0":
+            return "0 (1 limb)"
+
+        if val[0] == "-":
+            tmp = "negative"
+            val = val[1:]
+        else:
+            tmp = "positive"
+        if val[0] == "0":
+            tmp += " with leading zero limb"
+        elif len(val) > 10:
+            tmp = "large " + tmp
+        return tmp
+
+    @classmethod
+    def get_value_pairs(cls) -> Iterator[Tuple[str, ...]]:
```

Comment:
Isn't this, and `input_cases`, an iterator over pairs of strings? So `Iterator[Tuple[str, str]]`.

Reply:
This is correct, in these two cases mypy assumes the type `Iterator[Tuple[str, ...]]` even if the tuple is always of a fixed-width. I've added casts for these so that the correct type can be used.